network/p2p/cache/node_blocklist_wrapper.go

```diff
@@ -0,0 +1,191 @@
+// loaded from the data base (or assumed to be empty if no data base entry is present).
+func NewNodeBlocklistWrapper(identityProvider id.IdentityProvider, db *badger.DB) (*NodeBlocklistWrapper, error) {
+	blocklist, err := retrieveBlocklist(db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read set of blocked node IDs from data base: %w", err)
+	}
+
+	return &NodeBlocklistWrapper{
+		db:               db,
+		identityProvider: identityProvider,
+		blocklist:        blocklist,
+	}, nil
+}
+
+// Update sets the wrapper's internal set of blocked nodes to `blocklist`. Empty list and `nil`
+// (equivalent to empty list) are accepted inputs. To avoid legacy entries in the data base, this
+// function purges the entire data base entry if `blocklist` is empty.
+// This implementation is _eventually consistent_, where changes are written to the data base first
+// and then (non-atomically!) the in-memory set of blocked nodes is updated. This strongly
+// benefits performance and modularity. No errors are expected during normal operations.
+func (w *NodeBlocklistWrapper) Update(blocklist flow.IdentifierList) error {
+	b := blocklist.Lookup() // convert slice to
+	err := persistBlocklist(b, w.db)
+	if err != nil {
+		return fmt.Errorf("failed to persist set of blocked nodes to the data base: %w", err)
+	}
+
+	w.m.Lock()
+	w.blocklist = b
+	w.m.Unlock()
+	return nil
+}
+
+// ClearBlocklist purges the set of blocked node IDs. Convenience function
+// equivalent to w.Update(nil). No errors are expected during normal operations.
+func (w *NodeBlocklistWrapper) ClearBlocklist() error {
+	return w.Update(nil)
+}
+
+// GetBlocklist returns the set of blocked node IDs.
+func (w *NodeBlocklistWrapper) GetBlocklist() flow.IdentifierList {
+	w.m.RLock()
+	defer w.m.RUnlock()
+
+	identifiers := make(flow.IdentifierList, 0, len(w.blocklist))
+	for i := range w.blocklist {
+		identifiers = append(identifiers, i)
+	}
+	return identifiers
+}
+
+// Identities returns the full identities of _all_ nodes currently known to the
+// protocol that pass the provided filter. Caution, this includes ejected nodes.
+// Please check the `Ejected` flag in the returned identities (or provide a
+// filter for removing ejected nodes).
+func (w *NodeBlocklistWrapper) Identities(filter flow.IdentityFilter) flow.IdentityList {
+	identities := w.identityProvider.Identities(filter)
+	if len(identities) == 0 {
+		return identities
+	}
+
+	// Iterate over all returned identities and set ejected flag to true. We
+	// copy both the return slice and identities of blocked nodes to avoid
+	// any possibility of accidentally modifying the wrapped IdentityProvider
+	idtx := make(flow.IdentityList, 0, len(identities))
+	w.m.RLock()
+	for _, identity := range identities {
+		if w.blocklist.Includes(identity.NodeID) {
+			var i flow.Identity = *identity // shallow copy is sufficient, because `Ejected` flag is in top-level struct
+			i.Ejected = true
```

Comment:
Just to confirm, this is NOT an in-place mutation, right? So next time, when `Identities` is called, this Eject field is still false, before setting to true again, right?

Are there test cases that verify:
- if a Node is not originally Ejected, after it's blacklisted and then un-blacklisted, then `Ejected` should be false.
- if a Node is not originally Ejected (i.e. due to slashing), after it's blacklisted and then un-blacklisted, then `Ejected` should still be true.

Reply:
>  this is NOT an in-place mutation, right?

Exactly. Its a little buried in the [`NodeBlocklistWrapper` documentation](https://github.com/onflow/flow-go/blob/bd14684c0656b18aa4e22365d200bb94a7229fb7/network/p2p/cache/node_blocklist_wrapper.go#L26-L35): https://github.com/onflow/flow-go/blob/bd14684c0656b18aa4e22365d200bb94a7229fb7/network/p2p/cache/node_blocklist_wrapper.go#L28-L30

I have some checks (see [here](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L112-L113), [here](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L128-L129), and [here](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L151-L155)) in the existing test [`TestBlacklistedNode`](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L83-L84) that confirm that the wrapper does not write back into the memory returned by the underlying `IdentityProvider`. 

I have refactored those a bit to make it more apparent that we check this in the tests

Reply:
> * if a Node is not originally Ejected, after it's blacklisted and then un-blacklisted, then Ejected should be false.
> * if a Node is originally Ejected (i.e. due to slashing), after it's blacklisted and then un-blacklisted, then Ejected should still be true.

Also very good point. Added the test [`TestBlocklistAddRemove`](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L184-L236) that checks your scenarios.

Reply:
also added the tests
* [`TestUpdate`](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L243) for testing updating, clearing and retrieving the blocklist
* and [`TestDataBasePersist`](https://github.com/onflow/flow-go/blob/32c9d741db63f80c09f2dded083380b9e72f3fc3/network/p2p/cache/node_blocklist_wrapper_test.go#L266) which verifies that blocklist is persisted in database